From bdc4845cd844e674f17161435a11e60dde2769cc Mon Sep 17 00:00:00 2001
From: Jens Trillmann <jens.trillmann@governikus.de>
Date: Wed, 5 Jul 2023 09:33:03 +0200
Subject: Improve Intent source app detection

Activity.getReferrer does not only return app IDs but also URLs if
Intent.EXTRA_REFERRER is set on the Intent. In the case of Chrome the referrer
is set to the website triggering the Intent. To improve the detection of the
calling app we check first if the browser specific
Browser.EXTRAS_APPLICATION_ID is set. If it is not set we fall back to
Intent.getReferrer.

Pick-to: 6.6
Change-Id: I33d1edd52de98486d9616713e531ea20ada87bcb
---
 .../qtproject/qt/android/bindings/QtActivity.java | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git x/qtbase/src/android/java/src/org/qtproject/qt/android/bindings/QtActivity.java y/qtbase/src/android/java/src/org/qtproject/qt/android/bindings/QtActivity.java
index 9cef8146fd..f862f6aaee 100644
--- x/qtbase/src/android/java/src/org/qtproject/qt/android/bindings/QtActivity.java
+++ y/qtbase/src/android/java/src/org/qtproject/qt/android/bindings/QtActivity.java
@@ -16,6 +16,7 @@ import android.graphics.Canvas;
 import android.net.Uri;
 import android.os.Build;
 import android.os.Bundle;
+import android.provider.Browser;
 import android.util.AttributeSet;
 import android.view.ActionMode;
 import android.view.ActionMode.Callback;
@@ -237,9 +238,17 @@ public class QtActivity extends Activity
             return;
 
         String sourceInformation = "";
-        Uri referrer = getReferrer();
-        if (referrer != null)
-            sourceInformation = referrer.toString().replaceFirst("android-app://", "");
+        String browserApplicationId = intent.getExtras() == null ? "" : intent.getExtras().getString(Browser.EXTRA_APPLICATION_ID);
+        if (!browserApplicationId.isEmpty())
+        {
+            sourceInformation = browserApplicationId;
+        }
+        else
+        {
+            Uri referrer = getReferrer();
+            if (referrer != null)
+                sourceInformation = referrer.toString().replaceFirst("android-app://", "");
+        }
 
         intent.putExtra(EXTRA_SOURCE_INFO, sourceInformation);
     }
